packages/server/src/integrations/oracle.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  Integration,
+  DatasourceFieldTypes,
+  QueryTypes,
+  SqlQuery,
+} from "../definitions/datasource"
+import { getSqlQuery } from "./utils"
+import oracledb, { ExecuteOptions, Result,  Connection, ConnectionAttributes  } from "oracledb"
+import Sql from "./base/sql"
+
+module OracleModule {
+
+  oracledb.outFormat = oracledb.OUT_FORMAT_OBJECT;
+
+  interface OracleConfig {
+    host: string
+    port: number
+    database: string
+    user: string
+    password: string
+  }
+
+  const SCHEMA: Integration = {
+    docs: "https://github.com/oracle/node-oracledb",
+    friendlyName: "Oracle",
+    description: "Oracle Database is an object-relational database management system developed by Oracle Corporation",
+    datasource: {
+      host: {
+        type: DatasourceFieldTypes.STRING,
+        default: "localhost",
+        required: true,
+      },
+      port: {
+        type: DatasourceFieldTypes.NUMBER,
+        required: true,
+        default: 1521,
+      },
+      database: {
+        type: DatasourceFieldTypes.STRING,
+        required: true,
+      },
+      user: {
+        type: DatasourceFieldTypes.STRING,
+        required: true,
+      },
+      password: {
+        type: DatasourceFieldTypes.PASSWORD,
+        required: true,
+      }
+    },
+    query: {
+      create: {
+        type: QueryTypes.SQL,
+      },
+      read: {
+        type: QueryTypes.SQL,
+      },
+      update: {
+        type: QueryTypes.SQL,
+      },
+      delete: {
+        type: QueryTypes.SQL,
+      },
+    },
+  }
+  class OracleIntegration extends Sql {
+
+    private readonly config: OracleConfig
+
+    constructor(config: OracleConfig) {
+      super("oracle")
+      this.config = config
+    }
+
+    private query = async (query: SqlQuery): Promise<Result<any>> => { 
+      let connection
+      try {
+        connection = await this.getConnection()
```

Comment:
According to the oracle-node docs for `getConnection`:

> Obtains a connection from a pool in the connection pool cache or creates a new, standalone, non-pooled connection.

It doesn't seem like we have a pool here, but other data sources tend to use a pooled approach. We had some problems in the past with lots of connections being created, and pooling made things a bit more efficient. 

We can still use `getConnection`, but on the pool object rather than `oracledb` directly.
https://oracle.github.io/node-oracledb/doc/api.html#connpoolcache

Reply:
I'm not 100% sure if the pooled approach translates well to our application. If we were a traditional web app that connects to one database it would make a lot of sense, for us it's not as straightforward as we have many connection configurations. Normally you will create your pool once as a part of application initialization as early as possible. We would need to maintain a pool per tenant which would vastly increase the number of connections. Oracle actually handles this quite well as they provide a pool cache layer behind an alias as a part their api, but I'm not sure it's right for this use case and is probably intended for managing a small number of pools.

There's a part in the docs about the correlation between the connection pool sizes and nodes own thread pool https://oracle.github.io/node-oracledb/doc/api.html#-162-connections-threads-and-parallelism that tells us:
> The thread pool size should be equal to, or greater than, the maximum number of connections.

This sounds like maintaining lots connections can run into quite a bit of bother, I'd be more confident in the fire and forget method that we use currently. 

I had a look at the postgres integration for a reference on how we do pooling there, I'm not sure it's functioning as expected. Given the pool is `static`, I expected a global variable being overwritten for every new config being queried which sounds like there could be a lot of contention. When I debugged there I see that a new pool is created for every query being made and then thrown away e.g.

https://user-images.githubusercontent.com/8755148/141772894-72753971-b7f2-4f54-bb62-20de8d5730cb.mov

If we wanted to support pooling we could consider doing some dependency injection instead of creating the pool internally each time, but it's probably something we only want to do in single tenant mode

Reply:
> We would need to maintain a pool per tenant 

It might actually be a pool *per config, per tenant

Reply:
Nice one - great detail digging into that. Yeah it looks like it's "working"  for postgres in that it throws away the pool afterwards, which seems to give the same behaviour as fire and forget. Probably fine for now but since fire and forget works we should probably just make them all like that for consistency.

Actually now that I think about it - a lot of this was done before we even supported multi tenancy or cloud, where self-hosted users were reporting that they were blowing their maximum number of connections from using bb against a single database.

The key thing is that we just don't blow the database quota for connections, so think we are good here 👍 .